TEPHRA-210 Get table specific properties from tableDescriptor

gokulavasan opened the pull request:
And other properties from getConfiguration method, construct compactionState lazily

JIRA : https://issues.apache.org/jira/browse/TEPHRA-210

poornachandra commented on tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/coprocessor/TransactionProcessor.java:
```diff
@@ -140,29 +141,25 @@ public void start(CoprocessorEnvironment e) throws IOException {
-                                                                TxConstants.ALLOW_EMPTY_VALUES_DEFAULT);
+      String allowEmptyValuesFromTableDesc = tableDesc.getValue(TxConstants.ALLOW_EMPTY_VALUES_KEY);
+      this.allowEmptyValues = (allowEmptyValuesFromTableDesc != null) ? Boolean.valueOf(allowEmptyValuesFromTableDesc) :
+        getConfiguration(env).getBoolean(TxConstants.ALLOW_EMPTY_VALUES_KEY, TxConstants.ALLOW_EMPTY_VALUES_DEFAULT);
```
Shouldn't this property also be lazily read?

gokulavasan replied:
I think this property is used in preGetOp. Ideally, this should be read only from the tableDesc and set. If we do it lazily, then we might get different results for a Get operation based on whether we were able to read the Configuration successfully or not. I can modify this to read it only from tableDesc similar to the other property (READ_NON_TX_DATA)

poornachandra replied:
Looks like `allowEmptyValues` was not a table level property before. Any reason why we made it into a table level property now?

> If we do it lazily, then we might get different results for a Get operation based on whether we were able to read the Configuration successfully or not.

If not then it can be initialized incorrectly during startup, and will always be wrong, right? I think it should always be read lazily.

gokulavasan replied:
I have updated the PR to read the value from table descriptor first. If it is not available in the tableDescriptor, we try to obtain it from the configuration as we used to earlier.

poornachandra commented on tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/coprocessor/TransactionProcessor.java:
```diff
@@ -140,29 +141,25 @@ public void start(CoprocessorEnvironment e) throws IOException {
-        LOG.debug("Automatic invalid list pruning is enabled. Compaction state will be recorded in table " +
-                    pruneTable);
-      }
+        TimeUnit.SECONDS.toMillis(getConfiguration(env).getInt(TxConstants.Manager.CFG_TX_MAX_LIFETIME,
```
Same here about lazy read

gokulavasan replied:
This is being lazily read again when compaction is being called. Unfortunately we have to set it in the start method since it is in ensureValidTxLifetime method and thus we need to set it in the beginning.

poornachandra commented on tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/coprocessor/TransactionProcessor.java:
```diff
@@ -140,29 +141,25 @@ public void start(CoprocessorEnvironment e) throws IOException {
     }
   }
 
+  protected Configuration getConfiguration(CoprocessorEnvironment env) {
```
It would be good to add comments on the semantics of this method

gokulavasan replied:
Sure, will add comments for this.

gokulavasan commented:
@poornachandra Updated the PR by addressing comments. Please take a look again.

poornachandra commented on tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/coprocessor/TransactionProcessor.java:
```diff
@@ -303,10 +305,22 @@ public InternalScanner preCompactScannerOpen(ObserverContext<RegionCoprocessorEn
-    if (compactionState != null) {
+    if (pruneEnable == null) {
+      Configuration conf = getConfiguration(c.getEnvironment());
+      pruneEnable = conf.getBoolean(TxConstants.TransactionPruning.PRUNE_ENABLE,
```
If the configuration is not available for the first time then won't this become false for the whole run? Should we return null from `getConfiguration` in case configuration is not available? This way we can skip the run

gokulavasan replied:
In Tephra, it will always be available right?

poornachandra replied:
Since this code does not change between Tehpra and derived class implementation, we may need to handle the case when configuration is not available after startup. What do you think?

gokulavasan commented:
@poornachandra Addressed your comment by making getConfiguration method return a Configuration which could be nullable. Please look at ``ensureValidTxLifetime`` method since we now will have to throw an exception if Configuration is not available. Is that reasonable way to handle it?

poornachandra commented on tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/coprocessor/TransactionProcessor.java:
```diff
@@ -140,29 +143,37 @@ public void start(CoprocessorEnvironment e) throws IOException {
+        TxConstants.ALLOW_EMPTY_VALUES_DEFAULT;
+
+      // If the property is not present in the tableDescriptor, get it from the Configuration
+      this.allowEmptyValues = (allowEmptyValuesFromTableDesc != null) ?
```
It might help readability if we move reading `allowEmptyValues` into a private method

poornachandra commented:
LGTM

gokulavasan commented:
@poornachandra Please review the other commits. Ported changes to other compat modules.

poornachandra commented:
LGTM
